src/ui/ResultsPerPage/ResultsPerPage.ts

```diff
@@ -106,51 +110,87 @@ export class ResultsPerPage extends Component {
    */
   public setResultsPerPage(resultsPerPage: number, analyticCause: IAnalyticsActionCause = analyticsActionCauseList.pagerResize) {
     Assert.exists(resultsPerPage);
-    Assert.check(
-      this.options.choicesDisplayed.indexOf(resultsPerPage) != -1,
-      'The specified number of results is not available in the options.'
-    );
+    Assert.check(this.isValidChoice(resultsPerPage), 'The specified number of results is not available in the options.');
+
+    this.updateResultsPerPage(resultsPerPage);
+    this.updateQueryStateModelResultsPerPage();
+    this.logAnalyticsEvent(analyticCause);
+    this.executeQuery();
+  }
+
+  private updateResultsPerPage(resultsPerPage: number) {
     this.searchInterface.resultsPerPage = resultsPerPage;
     this.currentResultsPerPage = resultsPerPage;
+  }
+
+  private updateQueryStateModelResultsPerPage() {
+    this.queryStateModel.set(QueryStateModel.attributesEnum.rpp, this.currentResultsPerPage);
+  }
+
+  private logAnalyticsEvent(analyticCause: IAnalyticsActionCause) {
     this.usageAnalytics.logCustomEvent<IAnalyticsResultsPerPageMeta>(
       analyticCause,
       { currentResultsPerPage: this.currentResultsPerPage },
       this.element
     );
+  }
+
+  private executeQuery() {
     this.queryController.executeQuery({
       ignoreWarningSearchEvent: true,
       keepLastSearchUid: true,
       origin: this
     });
   }
 
+  private handleQueryStateModelChanged() {
+    const resultsPerPage = this.getInitialChoice();
+    this.updateResultsPerPage(resultsPerPage);
+  }
+
   private getInitialChoice(): number {
-    let initialChoice = this.options.choicesDisplayed[0];
-    if (this.options.initialChoice !== undefined) {
-      if (this.options.choicesDisplayed.indexOf(this.options.initialChoice) > -1) {
-        initialChoice = this.options.initialChoice;
-      } else {
-        this.logger.warn(
-          'The initial number of results is not within the choices displayed. Consider setting a value that can be selected. The first choice will be selected instead.'
-        );
+    const firstDisplayedChoice = this.options.choicesDisplayed[0];
+    const configuredChoice = this.options.initialChoice;
+    const queryStateModelChoice = this.queryStateModel.get(QueryStateModel.attributesEnum.rpp);
+    const queryStateModelChoiceIsNotDefault = queryStateModelChoice !== firstDisplayedChoice;
+
+    if (queryStateModelChoiceIsNotDefault && this.isValidChoice(queryStateModelChoice)) {
+      return queryStateModelChoice;
+    }
+
+    if (configuredChoice !== undefined) {
+      if (this.isValidChoice(configuredChoice)) {
+        return configuredChoice;
       }
+      this.logInvalidChoiceWarning();
     }
-    return initialChoice;
+
+    return firstDisplayedChoice;
+  }
+
+  private isValidChoice(choice: number) {
+    return this.options.choicesDisplayed.indexOf(choice) !== -1;
+  }
+
+  private logInvalidChoiceWarning() {
+    this.logger.warn(
```

Comment:
Could we add information about what the list of "valid choices" is , and what is the value we detected in the warning message ?